accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -869,33 +870,40 @@ public void complete()
 
     public static class Synchronized extends InMemoryCommandStore
     {
-        Runnable active = null;
+        private final AtomicBoolean running = new AtomicBoolean(false);
         final Queue<Runnable> queue = new ConcurrentLinkedQueue<>();
 
         public Synchronized(int id, NodeTimeService time, Agent agent, DataStore store, ProgressLog.Factory progressLogFactory, EpochUpdateHolder epochUpdateHolder)
         {
             super(id, time, agent, store, progressLogFactory, epochUpdateHolder);
         }
 
-        private synchronized void maybeRun()
+        private void maybeRun()
         {
-            if (active != null)
+            if (!running.compareAndSet(false, true))
                 return;
-
-            active = queue.poll();
-            while (active != null)
+            try
```

Comment:
TODO: add `synchronized` to keep old locking behavior; just make sure you acquire the lock first